refactor(core): Ensure registry and map are empty on view destroy

thePunderWoman opened the pull request:
This cleans up the memory usage of the defer block registry and jsactionmap when a view is destroyed that contains a defer block that is not yet hydrated.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->


- [x] Refactoring (no functional changes, no api changes)



## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

thePunderWoman committed:
refactor(core): Ensure registry and map are empty on view destroy

This cleans up the memory usage of the defer block registry and jsactionmap when a view is destroyed that contains a defer block that is not yet hydrated.

AndrewKushnir commented on packages/core/src/defer/instructions.ts:
```diff
@@ -1444,11 +1449,12 @@ export function triggerDeferBlock(lView: LView, tNode: TNode) {
 
   let registry: DeferBlockRegistry | null = null;
-  if (isIncrementalHydrationEnabled(injector)) {
+  if (isIncrementalHydrationEnabled(injector) && lDetails[SSR_UNIQUE_ID] !== null) {
     registry = injector.get(DEFER_BLOCK_REGISTRY);
+    registry.invokeCleanupFns(lDetails[SSR_UNIQUE_ID]);
```
nit: minor refactoring to avoid extra property reads:
```suggestion
  const ssrUniqueId = lDetails[SSR_UNIQUE_ID];
  if (isIncrementalHydrationEnabled(injector) && ssrUniqueId !== null) {
    registry = injector.get(DEFER_BLOCK_REGISTRY);
    registry.invokeCleanupFns(ssrUniqueId);
```
I'm also wondering if we should rename `registry.invokeCleanupFns` to `registry.invokeTriggerCleanupFns` (so that it's more aligned with the `invokeAllTriggerCleanupFns` function, which we can also rename to `invokeTriggerCleanupFns`)?

AndrewKushnir commented on packages/core/src/defer/registry.ts:
```diff
@@ -40,6 +43,13 @@ export class DeferBlockRegistry {
     this.registry.delete(blockId);
   }
+
+  cleanup(blockId: string) {
```
The `cleanup` and `remove` function names have relatively similar meaning in a context of this class, so it might be a bit confusing which once should be called at a particular code location. I'd propose leaving a TODO for now and do some refactoring in a followup PR.

thePunderWoman replied:
Actually, `remove` is only called in the registry twice. I'm just going to eliminate the remove function since it now is unnecessary.

AndrewKushnir commented on packages/core/src/event_delegation_utils.ts:
```diff
@@ -79,12 +80,16 @@ export const sharedMapFunction = (rEl: RElement, jsActionMap: Map<string, Set<El
       if (jsActionMap.has(blockName)) {
         blockList = [...blockList, ...jsActionMap.get(blockName)!];
         jsActionMap.delete(blockName);
+        if (registry !== null) {
+          registry.cleanup(blockName);
+        }
```
This call here is a bit unexpected. Looking at the call sites, I'm wondering if the `incrementallyHydrateFromBlockName` function only requires a call to cleanup a registry (so we can move that call there)? Other locations where the `removeListenersFromBlocks` function is called seem related to the event replay, but not incremental hydration specifically.

thePunderWoman replied:
I have a much cleaner version of this, but circular references are the problem.

AndrewKushnir commented on packages/core/src/defer/instructions.ts:
```diff
@@ -328,7 +328,12 @@ export function ɵɵdefer(
   }
 
-  const cleanupTriggersFn = () => invokeAllTriggerCleanupFns(lDetails, registry);
+  const cleanupTriggersFn = () => {
```
nit: since the function is now responsible for cleaning up the registry, we can rename the function:
```suggestion
  const onLViewDestroy = () => {
```

thePunderWoman committed:
refactor(core): clean up incremental code base

This re-organizes the code to be a bit cleaner with no more need for the onTriggerFn function.

thePunderWoman commented on packages/core/src/defer/instructions.ts:
```diff
@@ -1481,6 +1475,16 @@ export function triggerDeferBlock(lView: LView, tNode: TNode) {
   }
 }
 
+export async function hydrateFromBlockName(
```
Eliminating the `blocks.ts` file and moving calls to instructions removes the need for some of the circular dep problems we were having and centralizes the hydration code.

AndrewKushnir commented on packages/core/src/event_delegation_utils.ts:
```diff
@@ -79,12 +80,13 @@ export const sharedMapFunction = (rEl: RElement, jsActionMap: Map<string, Set<El
       if (jsActionMap.has(blockName)) {
         blockList = [...blockList, ...jsActionMap.get(blockName)!];
-        jsActionMap.delete(blockName);
+        registry.cleanup(blockName);
```
This feels a bit inverted: I guess we should try to call this function _from_ the registry and this particular function should not be aware of a registry at all. WDYT?

AndrewKushnir commented on packages/core/src/hydration/cleanup.ts:
```diff
@@ -129,3 +132,21 @@ export function cleanupDehydratedViews(appRef: ApplicationRef) {
   }
 }
+
+export async function cleanupDeferBlock(
```
I'd propose adding a JsDoc here to explain the role of this function.

AndrewKushnir commented on packages/core/src/hydration/cleanup.ts:
```diff
@@ -129,3 +132,21 @@ export function cleanupDehydratedViews(appRef: ApplicationRef) {
+    cleanupLContainer(deferBlock.lContainer);
+    cleanupDehydratedViews(injector.get(ApplicationRef));
+
+    // we need to wait for app stability here so we don't continue before
+    // the hydration process has finished, which could result in problems
+  }
+  return whenStable(injector.get(ApplicationRef));
```
nit:
```suggestion
  }
  // we need to wait for app stability here so we don't continue before
  // the hydration process has finished, which could result in problems
  return whenStable(injector.get(ApplicationRef));
```

AndrewKushnir commented on packages/core/src/hydration/event_replay.ts:
```diff
@@ -135,7 +132,9 @@ export function withEventReplay(): Provider[] {
                 initEventReplay(eventContractDetails, injector);
-                removeListenersFromBlocks([''], injector);
+                const jsActionMap = injector.get(JSACTION_BLOCK_ELEMENT_MAP);
+                jsActionMap.get('')?.forEach(removeListeners);
```
As an idea: I think it'd be great to introduce a const, which would help with code readability, for example:
```suggestion
                jsActionMap.get(EAGER_CONTENT_LISTENERS_KEY)?.forEach(removeListeners);
```

AndrewKushnir reviewed (commented):
@thePunderWoman great refactoring! 👍

thePunderWoman commented on packages/core/src/defer/instructions.ts:
```diff
@@ -1443,12 +1436,13 @@ export function triggerDeferBlock(lView: LView, tNode: TNode) {
+  let registry: DehydratedBlockRegistry | null = null;
+  if (isIncrementalHydrationEnabled(injector) && lDetails[SSR_UNIQUE_ID] !== null) {
+    registry = injector.get(DEHYDRATED_BLOCK_REGISTRY);
+    registry.invokeTriggerCleanupFns(lDetails[SSR_UNIQUE_ID]);
```
As a quick test, I tried it, but it breaks tests. This happens too early in the process to fully clean up all of the references. So alternatively, if we could move this specific call to later in the process, then possibly yes.

AndrewKushnir commented on packages/core/src/defer/instructions.ts:
```diff
@@ -552,14 +561,10 @@ export function ɵɵdeferHydrateOnImmediate() {
-        lDetails[SSR_UNIQUE_ID]!,
-        (deferBlock: DeferBlock) => triggerAndWaitForCompletion(deferBlock),
-      );
+      const lDetails = getLDeferBlockDetails(lView, tNode);
```
It looks like we already have an `lDetails` reference, so we can drop this line:
```suggestion
```

AndrewKushnir commented on packages/core/src/defer/instructions.ts:
```diff
@@ -891,15 +896,10 @@ export function scheduleDelayedHydrating(
-      ),
-    injector,
-  );
+  const ssrUniqueId = lDetails[SSR_UNIQUE_ID]!;
```
```suggestion
```

thePunderWoman replied:
This doesn't seem correct given the usages of ssrUniqueId on the following two lines. Am I missing something?

AndrewKushnir replied:
You are right 👍

thePunderWoman commented:
This PR was merged into the repository by commit 47224fb3c1b6885a6460615186b2c82cf4c5a596.

The changes were merged into the following branches: main, 19.0.x
